Fix the inconsistent formatting of the numeric columns in the Leaderboards on the Analytics > Dashboard page

eason9487 opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Closes #36748

This PR fixes the issue by:

- Add the `format` property to the items of the `rows` array in Analytics' Leaderboards API response to indicate how it's formatted.
- Make the `Leaderboard` component within Analytics aware of the currency context and format the numeric columns accordingly.

Extra changes to the *client/analytics/components/leaderboard/test/index.js* file:

- Apply better practice
- Avoid a warning when running JS tests
   ![3](https://github.com/woocommerce/woocommerce/assets/17420811/87ea3cc2-f46d-4efe-9986-d9087d52fc4f)
   Ref: https://github.com/woocommerce/woocommerce/blob/9.0.2/packages/js/components/src/pagination/pagination.tsx#L44

### Screenshots:

#### 📷 Using store currency options

![1](https://github.com/woocommerce/woocommerce/assets/17420811/cab23874-80f2-4da4-aeff-cdf29e82e68e)

#### 📷 Using filtered currency options

![2](https://github.com/woocommerce/woocommerce/assets/17420811/5c07cbff-86bd-49d4-8661-9437d539a84c)

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Checkout this PR, and run `pnpm install` and `pnpm build`
2. Open another webpage to view the current currency options on the **WooCommerce > Settings > General** page
3. Go to the **Analytics > Overview** page
4. If there are not large enough values to test, modifying these two lines locally can adjust the API response
   - https://github.com/woocommerce/woocommerce/blob/45022b3705b3b5751870e09cd8603ea1e58942ea/plugins/woocommerce/src/Admin/API/Leaderboards.php#L203
   - https://github.com/woocommerce/woocommerce/blob/45022b3705b3b5751870e09cd8603ea1e58942ea/plugins/woocommerce/src/Admin/API/Leaderboards.php#L208
5. Check if the currency options are applied to the currency and formatting of all amount values in the Leaderboards
6. Check if the thousand separator of the currency options is applied to all values of the Items Sold column in the Leaderboards
7. Add the following PHP code snippet, where the currency options should be different from the ones in step 2:
   ```php
   add_action( 'admin_enqueue_scripts', function () {
   	wp_add_inline_script( 'wp-hooks', <<<EOD
   	wp.hooks.addFilter( 'woocommerce_admin_report_currency', 'dev-blog-example', function() {
   		return {
   			code: "NZD",
   			decimalSeparator: ".",
   			precision: 2,
   			priceFormat: "%1\$s %2\$s",
   			symbol: "\$NZ",
   			symbolPosition: "left",
   			thousandSeparator: ",",
   		};
   	} );
   	EOD );
   } );
   ```
8. Repeat the test similar to steps 5 and 6, but check that the formatting of the values reflects the currency options returned by the filter in step 7.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

eason9487 committed:
Add the `format` property to the items of the `rows` array in Analytics' Leaderboards API response to indicate how it's formatted.

eason9487 committed:
Make the `Leaderboard` component within Analytics aware of the currency context and format the numeric columns accordingly.

eason9487 committed:
Avoid a warning from the test of `Leaderboard` component when running JS tests.

File: client/analytics/components/leaderboard/test/index.js
Ref: https://github.com/woocommerce/woocommerce/blob/9.0.2/packages/js/components/src/pagination/pagination.tsx#L44

eason9487 committed:
Apply better practice to the test of `Leaderboard` component.

File: client/analytics/components/leaderboard/test/index.js

eason9487 committed:
Add changelog.

tomalec commented on plugins/woocommerce-admin/client/analytics/components/leaderboard/index.js:
```diff
@@ -38,18 +79,7 @@ export class Leaderboard extends Component {
-					),
-					value: column.value,
-				};
-			} );
+			return row.map( this.getFormattedColumn.bind( this ) );
```
This creates a new function for every `row`, I think we can make it so we'll create a new function for each `getFormattedRows` call:
```js
		const getFormattedColumn = this.getFormattedColumn.bind( this );

		return this.props.rows.map( ( row ) => {
			return row.map( getFormattedColumn );
```

eason9487 replied:
Thanks for the suggestion. Adjusted in https://github.com/woocommerce/woocommerce/pull/49097/commits/0cf5d717b9442251f19a24e8f630cd0ff69e2b3c.

tomalec reviewed (approved):
I tested the changes locally and reviewed the code, despite a small comment regarding function binding LGTM

eason9487 committed:
Make the `Leaderboard` component not create a new function for every `row` rendering.

Address: https://github.com/woocommerce/woocommerce/pull/49097#discussion_r1667009910

eason9487 committed:
Merge branch 'trunk' into fix/36748-analytics-dashboard-currency-formatting
